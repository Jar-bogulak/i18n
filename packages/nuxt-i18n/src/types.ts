import type { Directions, Strategies, LocaleObject, VueI18nRoutingOptions } from 'vue-i18n-routing'
import type { Locale } from '@intlify/vue-i18n-bridge'

// TODO: bring up from @nuxtjs/i18n type definition

export type RedirectOnOptions = 'all' | 'root' | 'no prefix'

export interface LazyOptions {
  skipNuxtState?: boolean
}

export interface DetectBrowserLanguageOptions {
  alwaysRedirect?: boolean
  cookieCrossOrigin?: boolean
  cookieDomain?: string | null
  cookieKey?: string
  cookieSecure?: boolean
  fallbackLocale?: Locale | null
  redirectOn?: RedirectOnOptions
  useCookie?: boolean
}

export interface RootRedirectOptions {
  path: string
  statusCode: number
}

export type NuxtI18nOptions = {
  // differentDomains?: boolean
  // onBeforeLanguageSwitch?: (
  //   oldLocale: string,
  //   newLocale: string,
  //   initialSetup: boolean,
  //   context: NuxtContext
  // ) => string | void
  // onLanguageSwitched?: (oldLocale: string, newLocale: string) => void
  // detectBrowserLanguage?: DetectBrowserLanguageOptions | false
  langDir?: string | null
  lazy?: boolean | LazyOptions
  pages?: {
    [key: string]:
      | false
      | {
          [key: string]: false | string
        }
  }
  // parsePages?: boolean
  // rootRedirect?: string | null | RootRedirectOptions
  // routesNameSeparator?: string
  // skipSettingLocaleOnNavigate?: boolean
  // sortRoutes?: boolean
  // strategy?: Strategies
  // vueI18n?: I18nOptions | string
  // vueI18nLoader?: boolean
  // vuex?: VuexOptions | false
} & Pick<
  VueI18nRoutingOptions,
  | 'baseUrl'
  | 'defaultDirection'
  | 'defaultLocale'
  | 'defaultLocaleRouteNameSuffix'
  | 'locales'
  | 'routesNameSeparator'
  | 'trailingSlash'
>
