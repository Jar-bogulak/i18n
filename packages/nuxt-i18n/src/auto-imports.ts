import { isNuxt2 } from '@nuxt/kit'

import type { Nuxt } from '@nuxt/schema'

export async function setupAutoImports(nuxt: Nuxt) {
  // TOOD: resolve type definitions
  // register auto imports
  /*
  nuxt.hook('autoImports:extend', autoImports => {
    ;['useI18nHead', 'getRouteBaseName', 'localePath', 'localeRoute', 'localeLocation', 'switchLocalePath'].forEach(
      key => {
        autoImports.push({
          name: key,
          as: key,
          from: 'vue-i18n-routing'
        })
      }
    )
    autoImports.push({
      name: 'useI18n',
      as: 'useI18n',
      from: isNuxt2(nuxt) ? 'vue-i18n-bridge' : 'vue-i18n'
    })
  })
  */
}
