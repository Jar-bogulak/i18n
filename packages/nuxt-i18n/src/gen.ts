import createDebug from 'debug'
import { isString, isRegExp, isFunction, isArray, isObject } from '@intlify/shared'
import { templateUtils } from '@nuxt/kit'
import { genImport } from 'knitwork'

import type { NuxtI18nOptions, LocaleInfo } from './types'

export type LoaderOptions = {
  localeCodes?: string[]
  localeInfo?: LocaleInfo[]
  nuxtI18nOptions?: NuxtI18nOptions
}

const debug = createDebug('@nuxtjs/i18n:gen')

export function generateLoaderOptions(options: LoaderOptions = {}) {
  // TODO: lazy loading local info

  // prettier-ignore
  const genCode = `${Object.entries(options).map(([rootKey, rootValue]) => {
    if (rootKey === 'nuxtI18nOptions') {
      return `export const ${rootKey} = Object({${Object.entries(rootValue).map(([key, value]) => {
        if (key === 'vueI18n') {
          return `${key}: ${isObject(value)
            ? toCode(value)
            : isString(value)
              ? `(context) => import(${toCode(value)}).then(r => (r.default || r)(context))`
              : `${toCode({})}`
          }`
        } else {
          return `${key}: ${toCode(value)}`
        }
      }).join(`,`)}})`
    } else if (rootKey === 'localeInfo') {
      const localeInfo = options.localeInfo || []
      const importMapper = new Map<string, string>()
      localeInfo.forEach(({ code }) => {
        importMapper.set(code, templateUtils.importName(`locale_${code}`))
      })
      return `${localeInfo.map(l => genImport(l.path, importMapper.get(l.code))).join(`\n`)}
export const messages = () => Promise.resolve(Object({${[...importMapper].map(i => `${templateUtils.serialize(i[0])}:${i[1]}`).join(`,`)}}))`
	  } else {
	    return `export const ${rootKey} = ${toCode(rootValue)}`
	  }
  }).join('\n')}`

  debug('generate code', genCode)
  return genCode
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function stringifyObj(obj: Record<string, any>): string {
  return `Object({${Object.entries(obj)
    .map(([key, value]) => `${JSON.stringify(key)}:${toCode(value)}`)
    .join(`,`)}})`
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function toCode(code: any): string {
  if (code === null) {
    return `null`
  }

  if (code === undefined) {
    return `undefined`
  }

  if (isString(code)) {
    return JSON.stringify(code)
  }

  if (isRegExp(code) && code.toString) {
    return code.toString()
  }

  if (isFunction(code) && code.toString) {
    return `(${code.toString()})`
  }

  if (isArray(code)) {
    return `[${code.map(c => toCode(c)).join(`,`)}]`
  }

  if (isObject(code)) {
    return stringifyObj(code)
  }

  return code + ``
}
