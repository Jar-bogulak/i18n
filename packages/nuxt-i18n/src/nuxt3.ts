import createDebug from 'debug'
import { resolve } from 'pathe'
import { resolveModule, addPluginTemplate, extendWebpackConfig } from '@nuxt/kit'
import { distDir } from './dirs'

import type { Nuxt } from '@nuxt/schema'
import type { NuxtI18nOptions } from './types'

const debug = createDebug('@nuxtjs/i18n:nuxt3')

export async function setupNuxt3(options: NuxtI18nOptions, nuxt: Nuxt) {
  // TODO:
  // Resolve vue-i18n
  nuxt.options.alias['vue-i18n'] = resolveModule('vue-i18n/dist/vue-i18n.esm-bundler.js', {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n')

  // Resolve vue-i18n-routing
  nuxt.options.alias['vue-i18n-routing'] = resolveModule('vue-i18n-routing/dist/vue-i18n-routing.es.js', {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n-routing')

  addPluginTemplate({
    filename: 'runtime/nuxt3.plugin.mjs',
    src: resolve(distDir, 'runtime/nuxt3.plugin.mjs')
  })
}
