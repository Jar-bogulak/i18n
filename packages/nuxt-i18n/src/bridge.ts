import createDebug from 'debug'
import { createRequire } from 'module'
import { resolve } from 'pathe'
import { resolveModule, addPluginTemplate, extendWebpackConfig } from '@nuxt/kit'
import { distDir } from './dirs'
import webpack from 'webpack'

import type { Nuxt } from '@nuxt/schema'
import type { NuxtI18nOptions } from './types'

const debug = createDebug('@nuxtjs/i18n:bridge')

export async function setupNuxtBridge(options: NuxtI18nOptions, nuxt: Nuxt) {
  const _require = createRequire(import.meta.url)

  // Resolve vue-i18n
  nuxt.options.alias['vue-i18n'] = resolveModule('vue-i18n-legacy/dist/vue-i18n.esm.js', {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n')

  // Resolve vue-i18n-routing
  nuxt.options.alias['vue-i18n-routing'] = resolveModule('vue-i18n-routing/dist/vue-i18n-routing.es.js', {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n-routing')

  // Resolve vue-i18n-bridge
  nuxt.options.alias['vue-i18n-bridge'] = resolveModule('vue-i18n-bridge/dist/vue-i18n-bridge.esm-bundler.js', {
    paths: nuxt.options.modulesDir
  })
  nuxt.options.build.transpile.push('vue-i18n-bridge')

  // // Resolve vue-router via `@intlify/vue-router-bridge`
  // nuxt.options.alias['vue-router'] = resolveModule('@intlify/vue-router-bridge/lib/index.mjs', {
  //   paths: nuxt.options.modulesDir
  // })
  // nuxt.options.build.transpile.push('vue-router')

  // // Resolve vue-i18n via `@intlify/vue-i18n-bridge`
  // nuxt.options.alias['vue-i18n'] = resolveModule('@intlify/vue-i18n-bridge/lib/index.mjs', {
  //   paths: nuxt.options.modulesDir
  // })
  // nuxt.options.build.transpile.push('vue-i18n')

  addPluginTemplate({
    filename: 'runtime/bridge.plugin.mjs',
    src: resolve(distDir, 'runtime/bridge.plugin.mjs')
  })

  extendWebpackConfig(config => {
    config.plugins!.push(
      new webpack.DefinePlugin({
        __VUE_I18N_FULL_INSTALL__: 'true',
        __VUE_I18N_LEGACY_API__: 'true',
        __INTLIFY_PROD_DEVTOOLS__: 'false'
      })
    )
  })
}
