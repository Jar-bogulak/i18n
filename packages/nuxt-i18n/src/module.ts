import createDebug from 'debug'
import { defineNuxtModule, isNuxt2, isNuxt3 } from '@nuxt/kit'
import { setupNuxtBridge } from './bridge'
import { setupNuxt3 } from './nuxt3'

import type { NuxtI18nOptions } from './types'

const debug = createDebug('@nuxtjs/i18n:module')

export default defineNuxtModule<NuxtI18nOptions>({
  meta: {
    name: '@nuxt/i18n',
    configKey: 'i18n'
  },
  defaults: {},
  async setup(options, nuxt) {
    // TODO: implementation here
    console.log('@nuxt/i18n setup!')
    debug('options', options)

    if (isNuxt2(nuxt)) {
      await setupNuxtBridge(options, nuxt)
    } else if (isNuxt3(nuxt)) {
      await setupNuxt3(options, nuxt)
    } else {
      // TODO:
    }
  }
})
