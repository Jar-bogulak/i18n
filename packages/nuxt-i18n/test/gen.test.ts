import { it, expect } from 'vitest'
import { generateLoaderOptions } from '../src/gen'

it('generateLoaderOptions', () => {
  expect(
    generateLoaderOptions({
      localeCodes: ['en', 'ja', 'fr'],
      localeInfo: [
        {
          code: 'en',
          path: './locales/en.json'
        }
      ],
      nuxtI18nOptions: {
        defaultLocale: 'en',
        vueI18n: {
          locale: 'en',
          fallbackLocale: 'en',
          messages: {
            en: { hello: 'Hello!' }
          }
        }
      }
    })
  ).toMatchSnapshot()
})
