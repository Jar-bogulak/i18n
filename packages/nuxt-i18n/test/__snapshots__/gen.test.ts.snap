// Vitest Snapshot v1

exports[`generateLoaderOptions 1`] = `
"export const localeCodes = [\\"en\\",\\"ja\\",\\"fr\\"]
import localeEn_564a1578 from \\"./locales/en.json\\";
export const messages = () => Promise.resolve(Object({\\"en\\":localeEn_564a1578}))
export const nuxtI18nOptions = Object({defaultLocale: \\"en\\",vueI18n: Object({\\"locale\\":\\"en\\",\\"fallbackLocale\\":\\"en\\",\\"messages\\":Object({\\"en\\":Object({\\"hello\\":\\"Hello!\\"})})})})"
`;
